### PR TITLE
Move towards being able to build for x86

### PR DESCRIPTION
**Pull request by tannergooding**
This is work towards https://github.com/dotnet/machinelearning/issues/97

It allows you to build using `build.cmd -buildArch=x86` and, if you manually set the correct testhost, it allows the tests to pass.

### CODE REVIEW

File: Directory.Build.props [hunk 1]
```diff
@@ -7,13 +7,15 @@
   <PropertyGroup>
     <Configuration Condition="'$(Configuration)'==''">Debug</Configuration>
     <Platform Condition="'$(Platform)'==''">AnyCPU</Platform>
-    <NativeTargetArchitecture Condition="'$(NativeTargetArchitecture)' == ''">x64</NativeTargetArchitecture>
+	<TargetArchitecture Condition="'$(TargetArchitecture)' == ''">x64</TargetArchitecture>
```
Comment (danmosemsft) [inline Directory.Build.props]: nit: extra indent?

File: src/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer.csproj [hunk 2]
```diff
@@ -12,7 +12,7 @@
     <ProjectReference Include="..\..\Microsoft.ML.TensorFlow\Microsoft.ML.TensorFlow.csproj" />
   </ItemGroup>
 
-  <ItemGroup>
+  <ItemGroup Condition="'$(NativeTargetArchitecture)' == 'x64'">
```
Comment (danmosemsft) [inline src/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer.csproj]: Maybe worth `<!-- TensorFlow is 64 bit only -->`

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs [hunk 3]
```diff
@@ -18,7 +19,7 @@ public NormalizerTests(ITestOutputHelper output) : base(output)
         {
         }
 
-        [Fact]
+        [ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))]
```
Comment (eerhardt) [inline test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: What about this test and the below test require x64? I don't see tensorflow or lightgbm in the tests. Are there other things that only work on x64?

File: test/Microsoft.ML.TestFramework/TestCommandBase.cs [hunk 4]
```diff
@@ -812,7 +812,7 @@ public void CommandCrossValidation()
             Done();
         }
 
-        [Fact]
+        [ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))]
```
Comment (eerhardt) [inline test/Microsoft.ML.TestFramework/TestCommandBase.cs]: Similar question here - what requires x64?

File: src/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer.csproj (hunk 2 above)
Comment (eerhardt) [reply src/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer.csproj]: Agreed. I also think it would be good to document on each `[ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))]` what about the test doesn't allow it to be run on x86 processes.

File: test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs (hunk 3 above)
Comment (tannergooding) [reply test/Microsoft.ML.Tests/Transformers/NormalizerTests.cs]: As mentioned here: https://github.com/dotnet/machinelearning/pull/1008#issuecomment-424145620, there are a few tests which have different outputs from the baseline.

These need further investigation and potentially fixes.

File: Directory.Build.props (hunk 1 above)
Comment (tannergooding) [reply Directory.Build.props]: Fixed.

File: src/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer.csproj (hunk 2 above)
Comment (tannergooding) [reply src/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer/Microsoft.ML.DnnAnalyzer.csproj]: Fixed.

File: Directory.Build.props [hunk 5]
```diff
@@ -7,13 +7,15 @@
...
-    <NativeTargetArchitecture Condition="'$(NativeTargetArchitecture)' == ''">x64</NativeTargetArchitecture>
+    <TargetArchitecture Condition="'$(TargetArchitecture)' == ''">x64</TargetArchitecture>
+    <NativeTargetArchitecture Condition="'$(NativeTargetArchitecture)' == ''">$(TargetArchitecture)</NativeTargetArchitecture>
     <PlatformConfig>$(Platform).$(Configuration)</PlatformConfig>
   </PropertyGroup>
 
   <PropertyGroup>
     <RestoreSources>
       https://api.nuget.org/v3/index.json;
+      https://dotnetfeed.blob.core.windows.net/dotnet-core/index.json;
```
Comment (eerhardt) [inline Directory.Build.props]: Why do we need this new feed?

File: test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj [hunk 6]
```diff
@@ -8,7 +8,7 @@
     <NoWarn>2003;$(NoWarn)</NoWarn>
     <PublicSign>false</PublicSign>
     <SourceLink></SourceLink>
-    <PlatformTarget>x64</PlatformTarget>
+    <PlatformTarget>$(TargetArchitecture)</PlatformTarget>
```
Comment (eerhardt) [inline test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj]: Can this just be left to the default Any CPU?

File: test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs [hunk 7]
```diff
@@ -57,7 +57,7 @@ public OnnxTransformTests(ITestOutputHelper output) : base(output)
         {
         }
 
-        [Fact]
+        [ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))] // x86 fails with "An attempt was made to load a program with an incorrect format."
```
Comment (eerhardt) [inline test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs]: I assume Onnx support (i.e. Microsoft.ML.Scoring) is 64-bit only.  @shmoradims @jignparm - can you confirm?

File: test/Microsoft.ML.Predictor.Tests/TestPipelineSweeper.cs [hunk 8]
```diff
@@ -7,6 +7,7 @@
 using Microsoft.ML.Runtime.EntryPoints.JsonUtils;
 using Microsoft.ML.Runtime.PipelineInference;
 using Newtonsoft.Json.Linq;
+using System;
```
Comment (eerhardt) [inline test/Microsoft.ML.Predictor.Tests/TestPipelineSweeper.cs]: necessary?

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs [hunk 9]
```diff
@@ -102,7 +102,7 @@ public IList<TestDataset> GetDatasetsForClassificationWeightingPredictorsTest()
         /// <summary>
         ///A test for binary classifiers
         ///</summary>
-        [Fact]
+        [ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))] // x86 output differs from Baseline
```
Comment (eerhardt) [inline test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: I think we should fix/adjust the baseline mechanism so tests like this can be executed. We shouldn't disable tests just because the baseline differs.

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs [hunk 10]
```diff
@@ -587,7 +587,7 @@ public void RankingLightGBMTest()
             Done();
         }
 
-        [Fact]
+        [ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))] // x86 fails with "Unknown command: 'train'; Format error at (83,3)-(83,4011): Illegal quoting"
```
Comment (eerhardt) [inline test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: This should be investigated and fixed.

File: test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj (hunk 6 above)
Comment (tannergooding) [reply test/Microsoft.ML.FSharp.Tests/Microsoft.ML.FSharp.Tests.fsproj]: No, removing it, leaving it empty, and setting it to `AnyCPU` causes the tests to fail on x64

File: Directory.Build.props (hunk 5 above)
Comment (tannergooding) [reply Directory.Build.props]: For `Microsoft.DotNet.XUnitExtensions`

File: test/Microsoft.ML.Predictor.Tests/TestPipelineSweeper.cs (hunk 8 above)
Comment (tannergooding) [reply test/Microsoft.ML.Predictor.Tests/TestPipelineSweeper.cs]: ah no, will remove.

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs (hunk 9 above)
Comment (tannergooding) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: I agree, but it requires more investigation to determine what differs and why. I saw a number of other tests that are already disabled due to baseline differences as well (for all targets), that also need investigation.

Comment (tannergooding) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: Looking at a handful of the differences, they generally seem to be minor differences in the computed result (for example: `1.4157174495771871` vs `1.4157174495771872`). This is likely fallout from alignment differences and algorithm differences for various `System.Math` (and other) functions.

The comparison logic, in general, likely needs an overhaul (to properly account for these differences cross-architecture and cross-platform).

Comment (tannergooding) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: CC. @GalOshri, @shauheen, @TomFinley, @Zruty0

**Comment by eerhardt** [comment]
We will need to update:

https://github.com/dotnet/machinelearning/blob/master/pkg/common/CommonPackage.props
and
https://github.com/dotnet/machinelearning/blob/master/pkg/Microsoft.ML/build/netstandard2.0/Microsoft.ML.targets#L16

so we can support x86 on .NET Framework.

File: test/Microsoft.ML.Predictor.Tests/TestPipelineSweeper.cs (hunk 8 above)
Comment (tannergooding) [reply test/Microsoft.ML.Predictor.Tests/TestPipelineSweeper.cs]: Fixed.

**Comment by tannergooding** [comment]
Updated. Significantly fewer tests disabled now. The ones that are still disabled depend on a 64-bit only native component or currently have results that don't match for `digitsOfPrecision: 4`.

**Comment by tannergooding** [comment]
@sfilipi, would you be able to give this a quick review?

As per above, there are still a few tests that are disabled due to having less than 4 matching significant digits. From a brief look, some of these are caused by bugs in the parser/formatter code for full framework that have been fixed (or are being fixed) in .NET Core.

**Comment by sfilipi** [comment]
Taking a look shortly!

---
In reply to: [428336304](https://github.com/dotnet/machinelearning/pull/1008#issuecomment-428336304) [](ancestors = 428336304)

File: test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs (hunk 7 above)
Comment (eerhardt) [reply test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs]: Was there a response to this feedback? I think this seems like a good idea.  Or perhaps even deriving a few fact attributes like `OnnxFact` or `TensorflowFact`, etc.

---
In reply to: [222758463](https://github.com/dotnet/machinelearning/pull/1008#discussion_r222758463) [](ancestors = 222758463)

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs (hunk 10 above)
Comment (eerhardt) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: Was there a follow up issue for this?

---
In reply to: [222758690](https://github.com/dotnet/machinelearning/pull/1008#discussion_r222758690) [](ancestors = 222758690)

Comment (tannergooding) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: I've not logged issues for anything that requires followup yet.

File: test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs (hunk 7 above)
Comment (tannergooding) [reply test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs]: I'm not sure its worth the added overhead to declare a new fact, which itself calls `Environment.Is64BitProcess`.

The current conditional fact requires no "new" code and clearly defines that it is only supported on 64-bit processes.

Comment (eerhardt) [reply test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs]: And when we try to add `Aarch64` (i.e. `arm64`)?

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs [hunk 11]
```diff
@@ -192,7 +192,7 @@ public void MulticlassSdcaTest()
         /// <summary>
         /// Multiclass Logistic Regression test with a tree featurizer.
         /// </summary>
-        [Fact]
+        [ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))] // x86 output differs from Baseline
```
Comment (sfilipi) [inline test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: >[ConditionalFact(typeof(Environment), nameof(Environment.Is64BitProcess))] // x86 output differs from Baseline [](start = 8, length = 110)

Let's add an issue to create baselines for those x86 tests, after checking that the difference is ok.

File: test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs [hunk 12]
```diff
@@ -117,10 +117,10 @@ public void ChangePointDetectionWithSeasonality()
                 while (enumerator.MoveNext() && index < expectedValues.Count)
                 {
                     row = enumerator.Current;
-                    Assert.Equal(expectedValues[index++], row.Change[0]);
-                    Assert.Equal(expectedValues[index++], row.Change[1]);
-                    Assert.Equal(expectedValues[index++], row.Change[2]);
-                    Assert.Equal(expectedValues[index++], row.Change[3]);
+                    Assert.Equal(expectedValues[index++], row.Change[0], precision: 7);
```
Comment (sfilipi) [inline test/Microsoft.ML.TimeSeries.Tests/TimeSeriesDirectApi.cs]: >precision: 7); [](start = 73, length = 14)

is this necessary? i thought 7 was our default?

File: test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs (hunk 7 above)
Comment (tannergooding) [reply test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs]: Then we will need to re-review any of the currently disabled tests and possibly change this based on whether they support ARM and/or ARM64.

Comment (eerhardt) [reply test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs]: There are multiple advantages to taking the above approach:

1. It is self-documenting, like @danmosemsft said.
2. It is centralized, so if it ever needs to be updated, only need to update a single spot.
3. We can add any rules we want into the method, where you can't do this in a single attribute.

The only disadvantage I see is the small amount of effort it is going to take to make this change.

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs (hunk 11 above)
Comment (tannergooding) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: https://github.com/dotnet/machinelearning/issues/1215

File: test/Microsoft.ML.Predictor.Tests/TestPredictors.cs (hunk 10 above)
Comment (tannergooding) [reply test/Microsoft.ML.Predictor.Tests/TestPredictors.cs]: https://github.com/dotnet/machinelearning/issues/1216

File: test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs (hunk 7 above)
Comment (tannergooding) [reply test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs]: @eerhardt, do you have any problem with logging a tracking issue to improve this?

I'd like to get this merged in order to unblock people and I don't think it's going to be as trivial as just defining some new attributes, since we will need to do something different for .NET Core (where we need to instead check the `S.R.InteropServices.Architecture` of the process)

Comment (eerhardt) [reply test/Microsoft.ML.OnnxTransformTest/OnnxTransformTests.cs]: that's fine with me.
